### PR TITLE
fixes #10690 - fixing possible race condition with creating operating…

### PR DESCRIPTION
**Pull request by jlsherrill**
…systems

### CODE REVIEW

**Comment by ehelms** [comment]
Can you explain this a little more? Why it may or may not happen?

**Comment by jlsherrill** [comment]
Whenever you do a lookup and then a create there presents an opportunity for a race condition, consider two threads:

Step | Thread A | Thread B
-------|------------ | ---------------
 1 | does_exist?(os)  |  sleeping 
2 | sleeping  | does_exist? 
3 |  create(os) | sleeping 
4 | sleeping |  create(os) 

So if two repos are syncing an hit this point at the same time, they could get into this condition.  This is similar to http://projects.theforeman.org/issues/8586

and this follows the recommended procedure laid out here:  http://www.rubydoc.info/docs/rails/ActiveRecord/Relation:find_or_create_by

however in this case we don't get a 'ActiveRecord::RecordNotUnique' exception due to validators on the model.

I i'm still not 100% sure that the original cause was a race condition, but it is the best guess I have.  You'd think that it would be very rare for two repo syncs to interfere like this, yet the person that hit the problem could always simply hit resume and it would continue without error, following the same exact code path.  So it smells very much like a race condition

**Comment by ehelms** [comment]
And this is not something Dynflow resource locking can help us with at the moment?

**Comment by jlsherrill** [comment]
@ehelms nope, locking is currently meant to prevent other actions from even starting.  But in this case this is a very small step of an action, and we wouldn't even have the information in order to prevent lock the 2nd task out until the first task is mostly finished.

**Comment by ehelms** [comment]
I think I know your answer, but going to ask anyway, this is only accounting for the 2 repositories case (vs. say 3 or 4 simultaneously syncing and hitting this point) ?

File: app/models/katello/concerns/redhat_extensions.rb [hunk 1]
```diff
@@ -17,10 +17,13 @@ def find_or_create_operating_system(distribution)
...
 
-          os = ::Redhat.where(:name => os_name, :major => major, :minor => minor).first
-          os = create_operating_system(os_name, major, minor) unless os
+          create_os = lambda { ::Redhat.where(:name => os_name, :major => major, :minor => minor).first_or_create! }
 
-          return os
+          begin
+            create_os.call
+          rescue ActiveRecord::RecordInvalid
+            create_os.call
```
Comment (ehelms) [inline app/models/katello/concerns/redhat_extensions.rb]: So why call this again? Reading the notes on this, if the exception is thrown, that implies that the record already exists so calling it again would just simply find it but we don't appear to be needing to capture the return from the find itself.

Comment (jlsherrill) [inline app/models/katello/concerns/redhat_extensions.rb]: I could just call the where part again, if you prefer that I can change it:

```
begin
  ::Redhat.where(:name => os_name, :major => major, :minor => minor).first_or_create!
rescue
  ::Redhat.where(:name => os_name, :major => major, :minor => minor)
end
```

the OS is returned from the function as the begin/rescue clause is the last statement.  If you'd prefer to save it to a variable I can, but we typically don't for if/else's at the end of a function i don't think?

**Comment by jlsherrill** [comment]
The 3 and 4 case should still handle it I believe As 3 of the 4 should fail and hit the rescue, while one should succeed.

File: app/models/katello/concerns/redhat_extensions.rb (hunk 1 above)
Comment (ehelms) [inline app/models/katello/concerns/redhat_extensions.rb]: Ahhh, the return from before being removed is what threw me off.

**Comment by ehelms** [comment]
ACK

**Pull request merged**